File: leetcode/1299-replace-elements-with-greatest-element-on-right-side/ReplaceElementsWithGreatestElementOnRightSide.java
class Solution {
  public int[] replaceElements(int[] arr) {
    int[] result = new int[arr.length];
    result[arr.length - 1] = -1;

    int runningMax = -1;
    for (int i = arr.length - 2; i >= 0; i--) {
      runningMax = Math.max(runningMax, arr[i + 1]);
      result[i] = runningMax;
    }

    return result;
  }
}

File: leetcode/1299-replace-elements-with-greatest-element-on-right-side/replace-elements-with-greatest-element-on-right-side.ts
function replaceElements(arr: number[]): number[] {
  let max = -1;
  for (let i = arr.length - 1; i >= 0; i--) {
    let tmp = arr[i];
    arr[i] = max;
    max = Math.max(tmp, max);
  }

  return arr;
}
